src/ispc/kernels/lanczos3.ispc

```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
+        sum += weights[i];
+    }
+
+    // Normalize the weights, such that their sum is 1.0f
+    foreach(i = 0 ... end - start + 1) {
+        weights[i] /= sum;
+    }
 }
 
-static inline float byte_to_float(uint b) {
-    //return floatbits(0x3f800000 | (b << (23 - 8))) - 1.0;
-    return (float)b;
+struct WeightCollection {
+    uniform const uint32* starts;
+    uniform const uint32* weight_counts;
+    uniform const float* const* values;
+};
+
+struct Cache {
+    uniform WeightCollection vertical_weights;
+    uniform WeightCollection horizontal_weights;
+};
+
+uint8<3> sample_3_channels(varying uint64 read_address, const uniform uint8* varying src) {
+    // Memory reinterpretation to read all channels at once rather than one-by-one.
+    // While testing, this proved more performant than reading one-by-one.
+    const uniform uint8* pixel_ptr = src + read_address;
+    const uniform uint8<3>* pixel_ptr3 = (const uniform uint8<3>*)(pixel_ptr);
+    varying uint8<3> dst = {0, 0, 0};
+    dst = *pixel_ptr3;
+    return dst;
 }
 
-static inline uint8<4> resample_internal(uniform Image src_image, float<2> uv, uniform float<2> target_size, uniform float<2> inv_target_size, uniform uint8 num_channels) {
-    float<2> center_pixel = uv * target_size;
-    center_pixel.x = frac(center_pixel.x);
-    center_pixel.y = frac(center_pixel.y);
-    float<2> center = uv - (center_pixel - 0.5) * inv_target_size;
-    float<2> offset = (uv - center) * target_size;
-    const float inv_255 = rcp(255.0);
-
-    float<4> col = { 0, 0, 0, 0 };
-    float weight = 0.0;
-
-    for(uniform int x = -3; x < 3; x++){
-        for(uniform int y = -3; y < 3; y++){
-
-            float wx = lanczos3_filter((uniform float)x - offset.x);
-            float wy = lanczos3_filter((uniform float)y - offset.y);
-            float w = wx * wy;
-            float<2> texel_offset = {x, y};
-            float<2> texel_uv = center + texel_offset * inv_target_size;
-            int<2> pixel_coord = uv_to_pixel_id(src_image.size, texel_uv);
-
-            int addr = (pixel_coord.x + pixel_coord.y * src_image.size.x) * num_channels;
-
-            float<4> texel;
-
-            if (num_channels == 3) {
-                texel.x = byte_to_float(src_image.data[addr + 0]) * inv_255;
-                texel.y = byte_to_float(src_image.data[addr + 1]) * inv_255;
-                texel.z = byte_to_float(src_image.data[addr + 2]) * inv_255;
-            } else if (num_channels == 4) {
-                texel.x = byte_to_float(src_image.data[addr + 0]) * inv_255;
-                texel.y = byte_to_float(src_image.data[addr + 1]) * inv_255;
-                texel.z = byte_to_float(src_image.data[addr + 2]) * inv_255;
-                texel.w = byte_to_float(src_image.data[addr + 3]) * inv_255;
-            }
-
-            col += w * texel;
-            weight += w;
+void clean_and_write_3_channels(varying float<3> color, varying uint64 write_address, uniform uint8* varying dst) {
```

Comment:
Here and below: `write_address` and `dst` aren't used separately: can you combine them into one argument and let the caller perform `scratch_space + scratch_write_address`?